src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/taglets/SnippetTaglet.java

```diff
@@ -357,18 +333,28 @@ private Content generateContent(Element holder, DocTree tag)
             }
         }
 
-        String lang = null;
+        String lang;
         AttributeTree langAttr = attributes.get("lang");
-        if (langAttr != null) {
+
+        if (langAttr != null) { // the lang attribute overrides everything else
             lang = stringValueOf(langAttr);
-        } else if (containsClass) {
+        } else if (inlineContent != null && externalContent == null) { // an inline snippet
             lang = "java";
-        } else if (containsFile) {
-            lang = languageFromFileName(fileObject.getName());
+        } else if (externalContent != null) { // an external or a hybrid snippet
+            if (containsClass) { // the class attribute means Java
+                lang = "java";
+            } else {
+                String fileName = Path.of(fileObject.toUri()).getFileName().toString();
```

Comment:
The use of `Path.of` is (generally) questionable in a world of file objects. This assumes that there is a file system provider which can interpret the URL, which may or may not be the case for user-provided file objects. It would be better to do:
1. use `fileObject.toUri().getPath()`  to get the path component of the URI
2. either here or in `languageFromFileName` incorporate a check for the lastIndexOf `/` in the URI path. You could also use a regex for something like `\.[^/.]+$`. meaning dot followed by characters other than dot or slash up to the end of the string.